controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,287 @@
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-logr/logr"
+	metal3api "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+)
+
+// HostFirmwareComponentsReconciler reconciles a HostFirmwareComponents object.
+type HostFirmwareComponentsReconciler struct {
+	client.Client
+	Log                logr.Logger
+	ProvisionerFactory provisioner.Factory
+}
+
+// rhfcInfo is used to simplify the pass or arguments.
+type rhfcInfo struct {
+	ctx    context.Context
+	log    logr.Logger
+	hfc    *metal3api.HostFirmwareComponents
+	bmh    *metal3api.BareMetalHost
+	events []corev1.Event
+}
+
+func (info *rhfcInfo) publishEvent(reason, message string) {
+	t := metav1.Now()
+	hfcEvent := corev1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: reason + "-",
+			Namespace:    info.hfc.ObjectMeta.Namespace,
+		},
+		InvolvedObject: corev1.ObjectReference{
+			Kind:       "HostFirmwareComponents",
+			Namespace:  info.hfc.Namespace,
+			Name:       info.hfc.Name,
+			UID:        info.hfc.UID,
+			APIVersion: metal3api.GroupVersion.String(),
+		},
+		Reason:  reason,
+		Message: message,
+		Source: corev1.EventSource{
+			Component: "metal3-hostfirmwarecomponents-controller",
+		},
+		FirstTimestamp:      t,
+		LastTimestamp:       t,
+		Count:               1,
+		Type:                corev1.EventTypeNormal,
+		ReportingController: "metal3.io/hostfirmwarecomponents-controller",
+	}
+
+	info.events = append(info.events, hfcEvent)
+}
+
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwarecomponents,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwarecomponents/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwarecomponents/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *HostFirmwareComponentsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	reqLogger := r.Log.WithValues("hostfirmwarecomponents", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Get the corresponding baremetalhost in this namespace, if one doesn't exist don't continue processing
+	bmh := &metal3api.BareMetalHost{}
+	err = r.Get(ctx, req.NamespacedName, bmh)
+	if err != nil {
+		reqLogger.Info("could not get baremetal host, not running hostfirmwarecomponents reconciler")
```

Comment:
I'd prefer this moved under the 404 check and change to Error including the actual err. Otherwise, it's completely lost.

Reply:
To line 108 under the check on `IsNotFound`.

Reply:
oh ok, I've moved L104 to be after the if IsNotFound.